samples/csharp_dotnetcore/80.skills-simple-bot-to-bot/SimpleRootBot/AdapterWithErrorHandler.cs

```diff
@@ -13,42 +18,96 @@ namespace Microsoft.BotBuilderSamples.SimpleRootBot
+            await turnContext.SendActivityAsync(errorMessage);
+
+            // Send a trace activity, which will be displayed in the Bot Framework Emulator
+            await turnContext.TraceActivityAsync("OnTurnError Trace", exception.ToString(), "https://www.botframework.com/schemas/error", "TurnError");
+        }
+
+        private async Task<InvokeResponse> endSkillConversationAsync(ITurnContext turnContext)
+        {
+            if (_conversationState == null || _skillClient == null || _skillsConfig == null)
             {
-                // Log any leaked exception from the application.
-                logger.LogError(exception, $"[OnTurnError] unhandled error : {exception.Message}");
+                return null;
+            }
 
-                // Send a message to the user
-                var errorMessageText = "The bot encountered an error or bug.";
-                var errorMessage = MessageFactory.Text(errorMessageText, errorMessageText, InputHints.IgnoringInput);
-                await turnContext.SendActivityAsync(errorMessage);
+            try
+            {
+                await _conversationState.SaveChangesAsync(turnContext, true);
 
-                errorMessageText = "To continue to run this bot, please fix the bot source code.";
-                errorMessage = MessageFactory.Text(errorMessageText, errorMessageText, InputHints.ExpectingInput);
-                await turnContext.SendActivityAsync(errorMessage);
+                // Inform the active skill that the conversation is ended so that it has
+                // a chance to clean up.
+                // Note: "activeSkillProperty" is set by the RooBot while messages are being
+                // forwarded to a Skill.
+                var activeSkill = await _conversationState.CreateProperty<BotFrameworkSkill>("activeSkillProperty").GetAsync(turnContext, () => null);
+                var botId = _configuration.GetSection(MicrosoftAppCredentials.MicrosoftAppIdKey)?.Value;
 
-                if (conversationState != null)
-                {
-                    try
-                    {
-                        // Delete the conversationState for the current conversation to prevent the
-                        // bot from getting stuck in a error-loop caused by being in a bad state.
-                        // ConversationState should be thought of as similar to "cookie-state" in a Web pages.
-                        await conversationState.DeleteAsync(turnContext);
-                    }
-                    catch (Exception ex)
-                    {
-                        logger.LogError(ex, $"Exception caught on attempting to Delete ConversationState : {ex}");
-                    }
-                }
+                var endOfConversation = Activity.CreateEndOfConversationActivity();
+                endOfConversation.Code = "RootSkillError";
+                endOfConversation.ApplyConversationReference(turnContext.Activity.GetConversationReference());
```

Comment:
I don't think we need to apply the conversation reference here. SkillClient will do it for you.

Reply:
I'll test.  On Python I had to otherwise the conversation was null and it threw an exception.